scripts/generate-changelog/generate-changelog.go

```diff
@@ -14,27 +14,188 @@
 package main
 
 import (
+	"bufio"
+	"bytes"
 	"flag"
+	"fmt"
 	"os"
+	"os/exec"
+	"strings"
+	"time"
 
-	"github.com/prometheus/promu/pkg/changelog"
 	"github.com/sirupsen/logrus"
 )
 
-func main() {
-	version := flag.String("version", "", "release version")
-	flag.Parse()
-	f, err := os.Open("CHANGELOG.md")
+const (
+	feature        = "FEATURE"
+	enhancement    = "ENHANCEMENT"
+	bugfix         = "BUGFIX"
+	breakingChange = "BREAKINGCHANGE"
+)
+
+// kind represents the type of change.
+type kind int
+
+const (
+	kindBreakingChange = iota
+	kindFeature
+	kindEnhancement
+	kindBugfix
+	KindToBeIgnored
+)
+
+func getStringInBetweenTwoString(str string, startS string, endS string) (result string, found bool) {
+	s := strings.Index(str, startS)
+	if s == -1 {
+		return result, false
+	}
+	newS := str[s+len(startS):]
+	e := strings.Index(newS, endS)
+	if e == -1 {
+		return result, false
+	}
+	result = newS[:e]
+	return result, true
+}
+
+func getPreviousTag() string {
+	previousVersion, err := exec.Command("git", "describe", "--abbrev=0").Output()
 	if err != nil {
-		logrus.Fatal(err)
+		logrus.WithError(err).Fatal("unable to get the latest tag")
 	}
-	defer f.Close()
-	entry, err := changelog.ReadEntry(f, *version)
+	return strings.TrimSpace(string(previousVersion))
+}
+
+func getGitLogs(previousVersion string) []string {
+	// nolint: gosec
+	gitLogs, err := exec.Command("git", "log", fmt.Sprintf("%s...HEAD", previousVersion), "--pretty=oneline", "--no-decorate").Output()
 	if err != nil {
-		logrus.WithError(err).Fatalf("unable to get the entry in the changelog for the version %q", *version)
+		logrus.WithError(err).Fatal("unable to get the git logs")
+	}
+	entries := strings.Split(string(gitLogs), "\n")
+	if lastLine := entries[len(entries)-1]; strings.TrimSpace(lastLine) == "" {
+		entries = entries[0 : len(entries)-1]
+	}
+	return entries
+}
+
+func formatChangelogCategory(category string) string {
+	return fmt.Sprintf("[%s]", category)
+}
+
+// parseCatalogEntry returns the catalog kind and the catalog entry found.
+func parseCatalogEntry(entry string) (kind, string) {
```

Comment:
Since we are going to manually review the changelog for errors, I wonder if it would be useful to do something a little more complex here to make sure we don't accidentally drop features from the changelog because someone forgot to tag a commit with a catalog name, especially since this is new and humans are error-prone. 

What do you think about adjusting this to tag commits that we are unsure about with something like `[UNKNOWN]` and putting them at the bottom of the list for the releaser to manually review? I think we'd only want to do this if we also ignore commits that:
- We know we do not care about in the changelog:
  - Merge commits (e.g. starts with "Merge branch " or "Merge pull request ")
  - Release commits (e.g. starts with "Release " or "Sync Release ")
- Have an intentional "ignore" catalog name (e.g. `[IGNORE]`), so a committer can explicitly call out something is not intended for the changelog.

Maybe the result would look something like this? I also like the idea of putting in comments for the reviewer, so they know to review things. Deleting those comments when putting this up for a pull request would communicate they did the review.

```markdown
## 0.20.1 / 2022-12-08

<!-- CATEGORIZED ENTRIES. RELEASER, REVIEW AND ADJUST, AS NEEDED. -->

- [FEATURE] Discard Changes Confirmation Dialog (#834)
- [ENHANCEMENT] visual options and reset btn ux feedback (#850)
- [ENHANCEMENT] legend options editor UX improvements (#845)
- [ENHANCEMENT] Make it possible to adjust the height of the time range controls (#829)
- [ENHANCEMENT] Variable UX fixes (#842)
- [BREAKINGCHANGE] legend.position now required in time series panel (#848)

<!-- UNKNOWN ENTRIES. RELEASER, REVIEW AND CATALOG OR REMOVE, AS NEEDED. -->
- [UNKNOWN] Use 24hr time formatting in tooltip (#840)
- [UNKNOWN] Introduce TimeZoneProvider for specifying a timezone other than browser/local (#825)
```

Reply:
done. I added a list of exceptions that would make the commit ignore like suggested. The rest of the commit that doesn't contain any catalog entry will go the unknown list.